locator/tablets.cc

```diff
@@ -245,9 +245,48 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
     }
 }
 
-host_id tablet_map::get_primary_replica(tablet_id id) const {
-    const auto info = get_tablet_info(id);
-    return info.replicas.at(size_t(id) % info.replicas.size()).host;
+future<tablet_map::primary_replica_map> tablet_map::calc_primary_replica_map(table_id table, const topology& topo, sstring dc) const {
+    struct node_load {
+        uint64_t value = 0;
+    };
+    std::unordered_map<host_id, node_load> load_map;
+    primary_replica_map res;
+    res.reserve(tablet_count());
+    size_t empty_dcs = 0;
+    std::function<tablet_replica_set(const tablet_replica_set&)> filter_replicas;
+    if (dc.empty()) {
+        filter_replicas = [] (const tablet_replica_set& replicas) { return replicas; };
+    } else {
+        filter_replicas = [&] (const tablet_replica_set& replicas) {
+            return boost::copy_range<tablet_replica_set>(replicas | boost::adaptors::filtered([&] (const auto& tr) {
+                const auto* node = topo.find_node(tr.host);
+                return node && node->dc_rack().dc == dc;
+            }));
+        };
+    }
+    co_await for_each_tablet([&] (tablet_id tid, const tablet_info& ti) {
+        auto replicas = filter_replicas(ti.replicas);
+        if (replicas.empty()) {
+            if (dc.empty()) {
+                throw std::runtime_error(format("Could not find any replicas for table={} tablet={}", table, tid));
+            } else {
+                // can happen if rf=0 in the dc
+                ++empty_dcs;
+                return make_ready_future<>();
+            }
+        }
+        std::ranges::sort(replicas, std::less{}, [&] (const tablet_replica& tr) {
```

Comment:
I think we need std::ranges::stable_sort. The load_map of two nodes could be equal and the order of nodes matters.

Reply:
since the sort algorithm is symmetrical andcall nodes have the same starting conditions, they should all reach the same result, but maybe we should count on that. Some nodes may run a different version of the stl, for example.
stable_sort is one option, or alternatively the sort function can compare the host_id if the load is equal between any 2 replicas.

Reply:
I am not sure if the same stl version would give exact the same result on different runs, e.g.,  when the memory situation differs, or when it starts to switch between quick sort or heap sort under the hood. 

Most of the time stable_sort is also n*log(n).  Debugging bug due to unexpected sort would be a nightmare. Using stable_sort makes more sense.

Reply:
changed to stable_sort in v12